internal/r8/map_fetcher.go

```diff
@@ -0,0 +1,100 @@
+package r8
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const mapIndex = ".apm-android-map"
+
+var errFetcherUnvailable = errors.New("fetcher unavailable")
+
+type MapFetcher struct {
+	client *elasticsearch.Client
+	logger *logp.Logger
+}
+
+// NewMapFetcher returns a MapFetcher.
+func NewMapFetcher(c *elasticsearch.Client) *MapFetcher {
+	logger := logp.NewLogger(logs.Sourcemap)
+	return &MapFetcher{c, logger}
+}
+
+// Fetch fetches an R8 map from Elasticsearch.
+func (p *MapFetcher) Fetch(ctx context.Context, name, version string) ([]byte, error) {
+	resp, err := p.runSearchQuery(ctx, name, version)
+	if err != nil {
+		var networkErr net.Error
+		if errors.As(err, &networkErr) {
+			return nil, fmt.Errorf("failed to reach elasticsearch: %w: %v ", errFetcherUnvailable, err)
+		}
+		return nil, fmt.Errorf("failure querying ES: %w", err)
+	}
+
+	// handle error response
+	if resp.StatusCode >= http.StatusMultipleChoices {
+		defer resp.Body.Close()
```

Comment:
**issue**: we should move this before the if, right after the error check

Reply:
Thank you, I've just applied the changes.